### ISSUE TITLE
[RFC] side effect helper

### ISSUE THREAD

**Issue by fzaninotto** [marmelab/react-admin#2645]
## Problem

Adding side effects to actions from within a component is cumbersome. There is [the `callback` saga](https://marmelab.com/react-admin/Actions.html#custom-sagas), which allows to add a function to be executed on success or failure, but that requires a custom action. It can't be used e.g. after a regular crud action.

## Proposed solution

Provide a helper to add success or failure side effects to an existing action. The helper is an action creator that takes an action as parameter, and returns another action with side effects added.

Except I can't find a nice API with no drawbacks...

### Option 1 - jQuery style

* Pros: Explicit, familiar to frontend devs
* Cons: Old school

```jsx
import React from 'react';
import { withSideEffect, crudUpdate } from 'react-admin';
import { connect } from 'react-redux';

class ApproveButton extends React.Component {
    handleClick() {
        const { withSideEffect, id } = this.props;
        const action = crudUpdate('comment', id, { is_approved: true } );
        withSideEffect(action, {
            success: data => {
                // do something with the data
            },
            error: error => {
                // do something with the error
            }
        })
    }

    render() {
        // ...
    }
}

export default connect(null, { withSideEffect })(MyComponent);
```

### Option 2 - NodeJs style

Expect a callback as second argument.

* Pros: Familiar to NodeJs users
* Cons: This syntax is more uncommon on the frontend, and more and more deprecated on the backend. Also, it doesn't encourage catching errors. 

```jsx
import React from 'react';
import { withSideEffect, crudUpdate } from 'react-admin';
import { connect } from 'react-redux';

class ApproveButton extends React.Component {
    handleClick() {
        const { withSideEffect, id } = this.props;
        const action = crudUpdate('comment', id, { is_approved: true } );
        withSideEffect(action, (data, error) => {
            if (error) {
                // do something with the error
            }
            // do something with the data
        })
    }

    render() {
        // ...
    }
}

export default connect(null, { withSideEffect })(MyComponent);
```

### Option 3 - Positioned arguments style

* Pros: Concise
* Cons: Not explicit, not intuitive, requires reading the doc

```jsx
import React from 'react';
import { withSideEffect, crudUpdate } from 'react-admin';
import { connect } from 'react-redux';

class ApproveButton extends React.Component {
    handleClick() {
        const { withSideEffect, id } = this.props;
        const action = crudUpdate('comment', id, { is_approved: true } );
        withSideEffect(
            action,
            data => { // success side effect
                // do something with the data
            },
            error => { // error side effect
                // do something with the error
            })
    }

    render() {
        // ...
    }
}

export default connect(null, { withSideEffect })(MyComponent);
```

### Option 4 - fetch / Promise style

Requires treating actions as mutable objects, and modifying actions *after* being dispatched, but before being handled by sagas. Yes, it's possible.

* Pros: Familiar, Intuitive
* Cons: Misleading, error-prone (since this isn't a real Promise, so it can't be `await`ed nor chained further), Not robust (modifying actions after the reducer isn't a Redux documented feature)

```jsx
import React from 'react';
import { fetchWithRedux, crudUpdate } from 'react-admin';
import { connect } from 'react-redux';

class ApproveButton extends React.Component {
    handleClick() {
        const { fetchWithRedux, id } = this.props;
        const action = crudUpdate('comment', id, { is_approved: true } );
        fetchWithRedux(action)
            .then(data => {
                // do something with the data
            })
            .catch(error => {
                // do something with the error
            })
    }

    render() {
        // ...
    }
}

export default connect(null, { fetchWithRedux })(MyComponent);
```

### Option 5 - no style 

Same hack as option 4, except that, by renaming then and error to success and failure, it looks less like a Promise

* Pros: Somewhat familiar, although I don't remember which AJAX library used that syntax
* Cons: Still hackish and not robust

```jsx
import React from 'react';
import { withSideEffect, crudUpdate } from 'react-admin';
import { connect } from 'react-redux';

class ApproveButton extends React.Component {
    handleClick() {
        const { withSideEffect, id } = this.props;
        const action = crudUpdate('comment', id, { is_approved: true } );
        withSideEffect(action)
            .success(data => {
                // do something with the data
            })
            .failure(error => {
                // do something with the error
            })
    }

    render() {
        // ...
    }
}

export default connect(null, { withSideEffect })(MyComponent);
```

### Option 6 ? Feel free to suggest a better API!

**Comment by Kmaschta** [comment]
Your function reminds me of HTTP helpers [`on-finished`](https://www.npmjs.com/package/on-finished) and [`on-helpers`](https://www.npmjs.com/package/on-headers), so I would be in favor of the 2nd API (but with reverse argument positions):

```js
withSideEffect(action, (err, result) => {});
```

But to me, the real issue with this proposal is the name of the function, it isn't explicitf enough IMHO.
Instead, I would suggest more something like `dispatchWithSideEffect` or something like that.

Finally, isn't it simpler to let users install a redux middleware in order to catch whatever events they want?

**Comment by djhi** [comment]
> Finally, isn't it simpler to let users install a redux middleware in order to catch whatever events they want?

This is to avoid having to write code outside the component as it requires a lot of boilerplate to then retrieve the needed data in it

**Comment by Kmaschta** [comment]
Ok, I get it!

While you are at taking the API of React hooks, we can go further with:

```js
const [finished, error, result] = withSideEffect(action);
```
But it will be misleading if it isn't a hook.

**Comment by djhi** [comment]
I love it! Misleading but familiar with the hooks API. Besides, it might become a hook later :pray:

**Comment by fzaninotto** [comment]
@Kmaschta I don't understand your poposal. We're supposed to *create* success and error callbacks, not to *read* them from a hook. 

About the naming, we can't use ReactAdmin as it's not consistent with the rest of the functions we expose (which don't all contain "reactAdmin", thank God).

fetchResource isn't good either, because the fetching part is done by the action passed as parameter to this new creator. What it does is only to add success and error side effects.

**Comment by Kmaschta** [comment]
The expression I suggested is somewhat similar to the options 4 and 5: if the error or the result are defined, the user can take action.

But it would be easier to understand with another naming:

```js
const [error, response, request] = useAction(action);
// Error first force the user to handle it (Go-like !)
```

I must admit, my wording proposal wasn't as good as I hoped. My point was that we can use the hook synthax or the function should have a verb in it IMHO.

**Comment by Kmaschta** [comment]
Implemented by #2899 so I'm closing this one.
